System tests with HTML samples.

feerrenrut opened the pull request:
Example from "Fix double speaking label in angular checkbox" #10552 

### Summary of the issue:
- When fixing problems or introducing new features with NVDA working with a browser, we may accidentally introduce regressions. While working here we often develop small samples, but retesting these requires manual work (finding the samples, familiarizing with intent, using the sample correctly) which is prohibitive. 
- System tests written in the Robot Framework language are hard to read with a screen reader, and hard to write due to the reliance on speech output that often has (robot framework) argument separators contained in it.

### Description of how this pull request fixes the issue:
- Introduce automated tests that open Chrome with a specified sample move to the element, assert that the speech is as expected.
- Use Python for main system test logic. Tests still must have a Robot file, however now this just defines the metadata for the test and calls a single keyword which contains the test logic.
  - Test logic lives in python file with the same name as the robot file. 
- Libraries and test spy code has been re-arranged to make it more clear what is local, what is remote, and what library functions can be called. The convention for Robot Libraries to be in `/libraries` named `SomethingLib` the name of the module must match the name of the class used as the library, so the module is given an initial capital letter. Private functions / variables should all start with underscore. Public library "keywords" should use underscore to separate words.
- Test spy library calls are automatically wrapped with a method so that `run_keyword` no longer has to be called manually.

Overview of the refactoring done:
- `NVDA Core/startupShutdownNVDA.robot`
  - Logic moved to python: `robot/startupShutdownNVDA.py `
  - Metadata now in `robot/startupShutdownNVDA.robot`
- `libraries/nvdaRobotLib.py`
  - Mostly: `libraries/NvdaLib.py`
  - Code for installing the Spy and NVDA config. moved to `libraries/SystemTestSpy/configManager.py`
- `libraries/systemTestUtils.py` becomes `libraries/SystemTestSpy/blockUntilConditionMet.py`.
  - It is shared between Robot framework libraries and the SystemTestSpy package.
- `libraries/helperLib.py` code from `libraries/AssertsLib.py` 
- `libraries/sendKey.py` becomes `libraries/KeyInputLib.py` basically a straight rename.
- `libraries/speechSpy.py ` becomes `libraries/SystemTestSpy/speechSpySynthDriver.py`
- `libraries/systemTestSpy.py` becomes `libraries/SystemTestSpy/speechSpyGlobalPlugin.py`
- `NVDA Core/variables.py` no longer necessary, strings can easily be defined in test code now written in Python

The chrome tests are added:
- Logic in python: `robot/chromeTests.py `
- Metadata in `robot/chromeTests.robot`

### Testing performed:
Ran the tests locally and on Appveyor

### Known issues with pull request:
Doesn't close chrome at the completion of the test.

### Change log entry:

Section: Changes for developers:

`Automated testing of NVDA with Chrome and a HTML sample is now possible.`

feerrenrut committed:
Re-write system tests moving logic to Python.

Adds one basic chrome test.
Explicitly set Python version when running NVDA for system tests
as test runner does not inherit environment variables.

feerrenrut commented:
[The appveyor docs tell me](https://www.appveyor.com/docs/windows-images-software/#web-browsers) that Internet Explorer 11, two versions of Chrome, and two versions for Firefox are installed.

feerrenrut committed:
Fix chrome test on appveyor

feerrenrut committed:
More robust chrome setup

- Ignore separator in application title.
- Allow variations in the number of tab presses required to get to the start marker.

feerrenrut committed:
Fix lint errors

leonardder commented on tests/system/libraries/ChromeLib.py:
```diff
@@ -0,0 +1,134 @@
+
+		# move to start marker
+		for i in range(10):  # set a limit on the number of tries.
+			# Small changes in Chrome mean the number of tab presses to get into the document can vary.
```
have you considered pressing f6 to get to the document area? May be that's more reliable?

feerrenrut replied:
I hadn't considered that. For now I'll leave it as it is, but it's worth keeping in mind.

leonardder commented on tests/system/libraries/NvdaLib.py:
```diff
@@ -0,0 +1,243 @@
+	- NvdaLib.nvdaSpy is a library instance for getting speech and other information out of NVDA
+	"""
+	def __init__(self):
+		self.nvdaSpy = None  #: Optional[SystemTestSpy.speechSpyGlobalPlugin.NVDASpyLib]
```
Why is this annotated like this?

feerrenrut replied:
So that we don't have to import that module just for the type. That module depends on other bits of NVDA so it gets messy to import it.

LeonarddeR reviewed (commented):
I find this pretty hard to review, at it is not clear to me what's new code and what has been moved somehow. How would you advise to review this?

feerrenrut commented:
> I find this pretty hard to review, at it is not clear to me what's new code and what has been moved somehow. How would you advise to review this?

Yeah, I understand. The main thing is getting feedback to know if it's easy to understand how the tests are written. Would you be able to write a system test after looking at this for a bit?

A quick overview of the refactoring done:
- `NVDA Core/startupShutdownNVDA.robot`
  - Logic moved to python: `robot/startupShutdownNVDA.py `
  - Metadata now in `robot/startupShutdownNVDA.robot`
- `libraries/nvdaRobotLib.py`
  - Mostly: `libraries/NvdaLib.py`
  - Code for installing the Spy and NVDA config. moved to `libraries/SystemTestSpy/configManager.py`
- `libraries/systemTestUtils.py` becomes `libraries/SystemTestSpy/blockUntilConditionMet.py`.
  - It is shared between Robot framework libraries and the SystemTestSpy package.
- `libraries/helperLib.py` code from `libraries/AssertsLib.py` 
- `libraries/sendKey.py` becomes `libraries/KeyInputLib.py` basically a straight rename.
- `libraries/speechSpy.py ` becomes `libraries/SystemTestSpy/speechSpySynthDriver.py`
- `libraries/systemTestSpy.py` becomes `libraries/SystemTestSpy/speechSpyGlobalPlugin.py`
- `NVDA Core/variables.py` no longer necessary, strings can easily be defined in test code now written in Python

The chrome tests are added:
- Logic in python: `robot/chromeTests.py `
- Metadata in `robot/chromeTests.robot`

I have also updated the description of this PR to point out the conventions used:
> The convention for Robot Libraries to be in `/libraries` named `SomethingLib` the name of the module must match the name of the class used as the library, so the module is given an initial capital letter. Private functions / variables should all start with underscore. Public library "keywords" should use underscore to separate words.

feerrenrut commented:
Since mostly the code has only been re-organized, the most important thing is to ensure that the tests are easy to understand these are the files in `tests/system/robot/`

Other than that, one interesting change is how to interact with the remote library `NvdaSpy` (a global plugin that runs a server that the tests connect to in order to extract speech information from NVDA). Previously you had to call `run_keyword` passing in the function name as a string. Now, these are all converted to methods on the instance. With some type hint trickery you get intellisense (at least in pycharm, hopefully also in vsCode). You can see how this works in [libraries/NvdaLib.py _addMethodsToSpy](https://github.com/nvaccess/nvda/pull/10553/files#diff-adde7014c481b0cca4001ac351437756R171)

michaelDCurran reviewed (approved):
I'm happy with the changes, including being able to read the tests themselves. I would like however to see a second chrome test added before this is merged, so we have the confidence that multiple chrome tests don't cause strangeness with Chrome trying to be started twice or what ever would happen.

feerrenrut commented:
As we discussed, I have run the same test multiple times (leaving chrome open) which did not cause any issues. I'll go ahead and merge this.
